### PR TITLE
feat: typed Inputs and Outputs

### PR DESCRIPTION
**Pull request by J-Loudet**
This PR adds the possibility for users to associate a type to the Inputs and
Outputs of their Nodes.

Associating a type has two benefits:
1. On an Output, Rust will enforce that only an instance of the associated type
   is sent to downstream nodes,
2. On an Input, Zenoh-Flow will, transparently, try to "convert" the received
   data to a reference of the associated type.
   
Raw versions of the Input and Output, namely `InputRaw` and `OutputRaw` are
available to have non-typed variant.

The method `forward` was added to the `OutputRaw` for cases that require
performance: Zenoh-Flow will perform no verification and just forward the a
`LinkMessage` on the Output.

## Internal modifications

The `Inputs` and `Outputs` type aliases were rewritten in favor of concrete
structures. This allowed removing the (badly named) `Streams` traits.

The `Message` structure was renamed `LinkMessage`: it better describes how
it's used and where it "comes from".

The `Data` structure was renamed `Payload`.

The purpose of the last two modifications is to be able to present users with a
`Message` enumeration: it is either `Data<T>` or a `Watermark`.

Lastly, the synchronous methods `send` and `recv` were removed.

## Examples

``` rust
// `outputs` is an `Outputs` structure.
let output: Output<usize> = outputs.take("out").unwrap();
// alternative:
// let output = outputs.take::<usize>("out").unwrap();

// `inputs` is an `Inputs` structure.
let input: Input<usize> inputs.take("in").unwrap();
// alternative:
// let input = inputs.take::<usize>("out").unwrap();
```

### CODE REVIEW

File: zenoh-flow/src/io/input.rs [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+///
+/// To access these underlying channels, two methods are available:
+/// - `take`: this will return an `Input<T>` where `T` implements [`ZFData`](`ZFData`),
+/// - `take_raw`: this will return an [`InputRaw`](`InputRaw`) — a type agnostic receiver.
+///
+/// Choosing between `take` and `take_raw` is a trade-off between convenience and performance: an
+/// `Input<T>` conveniently receives instances of `T` and is thus less performant as Zenoh-Flow has
+/// to manipulate the data; an `InputRaw` is more performant but all the manipulation must be
+/// performed in the code of the Node. An `InputRaw` can be leveraged, for instance, for
+/// load-balancing or rate-limiting.
```
Comment (gabrik) [inline zenoh-flow/src/io/input.rs]: I would also add that `InputRaw` is used by bindings. As Python (or other bindings in the future) cannot understand types coming from Rust (and vice-versa)

File: zenoh-flow/src/io/input.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            .or_insert_with(Vec::default)
+            .push(rx)
+    }
+
+    /// Returns the typed [`Input<T>`](`Input`) associated to the provided `port_id`, if one is
+    /// associated, otherwise `None` is returned.
+    ///
+    /// ## Performance
+    ///
+    /// With a typed [`Input<T>`](`Input`), Zenoh-Flow will perform operations on the underlying
```
Comment (gabrik) [inline zenoh-flow/src/io/input.rs:72]: We could add that deserialization is done only if the inner data is serialized (i.e., coming from the network or the upstream node sent it as a byte array).

File: zenoh-flow/src/io/input.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            })
+    }
+}
+
+/// An [`InputRaw`](`InputRaw`) exposes the [`LinkMessage`](`LinkMessage`) it receives.
+///
+/// It's primary purpose is to ensure "optimal" performance. This can be useful to implement
+/// behaviour where actual access to the underlying data is irrelevant.
+#[derive(Clone, Debug)]
+pub struct InputRaw {
```
Comment (gabrik) [inline zenoh-flow/src/io/input.rs:116]: What happens if I use an `Output<T>` to send data to an `InputRaw`? 
My guess is that data need to be serialized by the `InputRaw::recv_async`, if not I guess that something bad will happen.

File: zenoh-flow/src/io/input.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    /// Returns the first [`LinkMessage`](`LinkMessage`) that was received, *asynchronously*, on any
+    /// of the channels associated with this Input.
+    ///
+    /// If several [`LinkMessage`](`LinkMessage`) are received at the same time, one is *randomly*
+    /// selected.
+    ///
+    /// ## Error
+    ///
+    /// If an error occurs on one of the channels, this error is returned.
+    pub async fn recv_async(&self) -> Result<LinkMessage> {
```
Comment (gabrik) [inline zenoh-flow/src/io/input.rs]: Are we going to add a `try_recv()` as briefly discussed?

File: zenoh-flow/src/io/input.rs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    /// - data received serialized is deserialized (an allocation is performed to store an instance
+    ///   of `T`),
+    /// - data received "typed" are checked against the type associated to this [`Input`](`Input`).
+    ///
+    /// ## Error
+    ///
+    /// Several errors can occur:
+    /// - one of the channels can return an error (e.g. a disconnection),
+    /// - Zenoh-Flow failed at interpreting the received data as an instance of `T`.
+    pub async fn recv_async(&self) -> Result<(Message<T>, Timestamp)> {
```
Comment (gabrik) [inline zenoh-flow/src/io/input.rs]: Are we going to add a `try_recv()` as briefly discussed?

File: zenoh-flow/src/io/output.rs [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    ///
+    /// If no `timestamp` is provided, the current timestamp — as per the [`HLC`](`HLC`) used by the
+    /// Zenoh-Flow daemon running this Node — is taken.
+    ///
+    /// ## Errors
+    ///
+    /// If an error occurs while sending the watermark on a channel, Zenoh-Flow still tries to send
+    /// it on the remaining channels. For each failing channel, an error is logged and counted for.
+    /// The total number of encountered errors is returned.
+    pub async fn send_async(&self, data: impl Into<Payload>, timestamp: Option<u64>) -> Result<()> {
```
Comment (gabrik) [inline zenoh-flow/src/io/output.rs]: Need a `try_send`, similar to `try_recv`

File: zenoh-flow/src/io/output.rs [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+    ///
+    /// Both `T` and `Data<T>` implement this constraint. Hence, in practice, any type that
+    /// implements `Into<T>` can be sent (provided that `Into::<T>::into(u)` is called first).
+    ///
+    /// ## Errors
+    ///
+    /// If an error occurs while sending the message on a channel, Zenoh-Flow still tries to send it
+    /// on the remaining channels. For each failing channel, an error is logged and counted for. The
+    /// total number of encountered errors is returned.
+    pub async fn send_async(&self, data: impl Into<Data<T>>, timestamp: Option<u64>) -> Result<()> {
```
Comment (gabrik) [inline zenoh-flow/src/io/output.rs]: Same as above

Comment (gabrik) [inline zenoh-flow/src/io/output.rs]: The PR description says that this function name should be `send`.

File: zenoh-flow/src/io/output.rs (hunk 6 above)
Comment (gabrik) [inline zenoh-flow/src/io/output.rs]: The PR description says that this function name should be `send`.

File: zenoh-flow/src/io/input.rs (hunk 4 above)
Comment (gabrik) [inline zenoh-flow/src/io/input.rs]: The PR description says that this function name should be `recv`.

File: zenoh-flow/src/io/input.rs (hunk 5 above)
Comment (gabrik) [inline zenoh-flow/src/io/input.rs]: The PR description says that this function name should be `recv`.

File: zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs [hunk 8]
```diff
@@ -30,7 +30,7 @@ use zenoh_util::core::AsyncResolve;
 /// different runtimes.
 pub(crate) struct ZenohSender {
     pub(crate) id: NodeId,
-    pub(crate) input: Input,
+    pub(crate) receivers: Vec<flume::Receiver<LinkMessage>>,
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs]: Why is not using the `Input`?

File: zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs [hunk 9]
```diff
@@ -87,26 +87,33 @@ impl Node for ZenohSender {
...
+            )
+            .into()),
         }
     }
 }
 
 /// A `ZenohReceiver` receives the messages from Zenoh when nodes are running on different runtimes.
 pub(crate) struct ZenohReceiver {
     pub(crate) id: NodeId,
-    pub(crate) output: Output,
```
Comment (gabrik) [inline zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs:109]: Why it is not using the `RawOutput`?

File: zenoh-flow/src/types/message.rs [hunk 10]
```diff
@@ -183,22 +314,169 @@ impl Message {
...
     fn eq(&self, other: &Self) -> bool {
         self.get_timestamp() == other.get_timestamp()
     }
 }
 
-impl Eq for Message {}
+impl Eq for LinkMessage {}
+
+/// A `Message<T>` is what is received on an `Input<T>`, typically after a call to `recv_async` or
+/// `recv`.
```
Comment (gabrik) [inline zenoh-flow/src/types/message.rs]: `recv` or `try_recv`

File: zenoh-flow/src/types/message.rs [hunk 11]
```diff
@@ -183,22 +314,169 @@ impl Message {
...
+///
+/// When deserializing, an allocation is performed.
+#[derive(Debug)]
+pub struct Data<T> {
+    payload: Payload,
+    bytes: Option<Vec<u8>>,
+    // CAVEAT: Setting the value **typed** to `None` when it was already set to `Some` can cause a
+    // panic! The visibility of the entire structure is private to force us not to make such
+    // mistake.
+    // In short, tread carefully!
```
Comment (gabrik) [inline zenoh-flow/src/types/message.rs:366]: Typo: treat

File: zenoh-flow/src/types/message.rs [hunk 12]
```diff
@@ -183,22 +314,169 @@ impl Message {
...
+    typed: Option<T>,
+}
+
+// CAVEAT: The implementation of `Deref` is what allows users to transparently manipulate the type
+// `T`. However, for things to go well, the `typed` field MUST BE MANIPULATED WITH CAUTION.
```
Comment (gabrik) [inline zenoh-flow/src/types/message.rs:371]: Add a Warning to the doc, not only the caps text

File: zenoh-flow/tests/dataflow.rs [hunk 13]
```diff
@@ -105,8 +107,8 @@ impl Node for GenericSink {
 // OPERATORS
 
 struct NoOp {
```
Comment (gabrik) [inline zenoh-flow/tests/dataflow.rs]: I would add an Operator with `Input` and `RawOutput` (and vice-versa) as well as a `Sink` with `RawInput` just to verify that all cases work.

**Review by gabrik — CHANGES_REQUESTED** [review]
Overall seems good, I guess there are still some cases that may cause troubles, as well as some things that need to be changed in function names (according to PR description) and in the `ZenohSender`/`ZenohReceiver` that should not manipulate directly the `flume::channel` but use the `Input`/`Output`.

File: zenoh-flow/src/io/input.rs (hunk 2 above)
Comment (J-Loudet) [reply zenoh-flow/src/io/input.rs:72]: The text in the line after isn't explicit enough?

> if the data are received serialized then Zenoh-Flow will deserialize them, if they are received "typed" then Zenoh-Flow will check that the type matches what is expected.

File: zenoh-flow/src/io/input.rs (hunk 3 above)
Comment (J-Loudet) [reply zenoh-flow/src/io/input.rs:116]: Nothing bad happens! :-)

An `Output<T>` will send a `Payload::Typed` to downstream nodes. Zenoh-Flow will thus serialize it (as `T` implements `ZFData`) only if need be.

File: zenoh-flow/src/io/input.rs (hunk 2 above)
Comment (gabrik) [reply zenoh-flow/src/io/input.rs:72]: I missed it :)

File: zenoh-flow/src/io/input.rs (hunk 4 above)
Comment (J-Loudet) [reply zenoh-flow/src/io/input.rs]: I did not want to "mix" too many changes in the same PR. As we want to remove the, potentially harmful, synchronous `recv` I saw no point of updating it to the changes and so removed it now. But this leaves things in an "half-done" state, so I'll add the `try_{recv|send}` and rename the methods `{recv|send}_async` to `{recv|send}`. :)

File: zenoh-flow/src/types/message.rs (hunk 11 above)
Comment (J-Loudet) [reply zenoh-flow/src/types/message.rs:366]: Although "treat" would indeed work, I did want to use the expression "tread carefully" (= be cautious) 😛

File: zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs (hunk 8 above)
Comment (J-Loudet) [reply zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs]: It's an oversight, I introduced the `InputRaw` and `OutputRaw` after making that change & forgot to port it

**Review by gabrik — APPROVED** [review]
LGTM!

**Pull request merged**